cfpb/cfgov-refresh
# Move test tooling to Jest

[Istanbul has been deprecated](https://github.com/gotwarlost/istanbul):

> Deprecation Notice: this version of istanbul is deprecated, we will not be landing pull requests or releasing new versions. But don't worry, the Istanbul 2.0 API is now available and is being actively developed in the new istanbuljs organization.

This is preventing some new tests from being written, such as those in https://github.com/cfpb/cfgov-refresh/pull/3830.

The Istanbul 2.0 API is significantly different. After chatting with @sebworks, instead of migrating istanbul, we could use [jest](https://facebook.github.io/jest/), which largely keeps the same syntax we have, but includes jsdom and istanbul code coverage.

## Additions

- Adds default Babel config to package.json.
- Adds Jest config to package.json.

## Removals

- Removes empty tests (Jest doesn't allow this).
- Removes jsdom from package.json
- Removes gulp-istanbul from package.json
- Removes mocha from package.json

## Changes

- Changes gulp test task to run Jest for JS unit tests.
- Changes `o-table-row-links.js` to set window URL with `window.location.assign(…)` from `window.location = …` so that the location method can be mocked in tests.
- Changes `o-table-row-links-spec.js` to check for calls to `window.location.assign(…)` instead of checks of the URL.
- Removes tests path from the gulp config in favor of retrieving the path through the environment.js.
- 

## Testing

1. Run `./frontend.sh`
2. Run `gulp unit:test` to run all JS unit tests.
3. Run `gulp test:unit --specs=js/modules/` to run a directory of specs.
3. Run `gulp test:unit --specs=js/modules/o-table-row-links-spec.js` to run a single spec.
4. Open `test/unit_test_coverage/lcov-report/index.html` to see code coverage report.

## Todos

- Chai and Sinon can be replaced with Jest too.
